sigoptlite/validators.py

```diff
@@ -0,0 +1,564 @@
+      raise ValueError(f"observation_budget is required for a {cls_name} with more than one optimized metric")
+    if experiment.has_constraint_metrics:
+      raise ValueError(f"observation_budget is required for a {cls_name} with constraint metrics")
+    if experiment.is_multitask:
+      raise ValueError(f"observation_budget is required for a {cls_name} with tasks (multitask)")
+
+  if not (experiment.optimized_metrics or experiment.constraint_metrics):
+    raise ValueError(f"{cls_name} must have optimized or constraint metrics")
+
+  if experiment.optimized_metrics:
+    if not len(experiment.optimized_metrics) in [1, 2]:
+      raise ValueError(f"{cls_name} must have one or two optimized metrics")
+    elif len(experiment.optimized_metrics) == 1 and experiment.optimized_metrics[0].threshold is not None:
+      raise ValueError(
+        "Thresholds are only supported for experiments with more than one optimized metric."
+        " Try an All-Constraint experiment instead by setting `strategy` to `constraint`."
+      )
+
+  # Check feature viability of multisolution experiments
+  num_solutions = experiment.num_solutions
+  if num_solutions and num_solutions > 1:
+    if num_solutions > observation_budget:
+      raise ValueError("observation_budget needs to be larger than the number of solutions")
+    if not len(experiment.optimized_metrics) == 1:
+      raise ValueError(f"{cls_name} with multiple solutions require exactly one optimized metric")
+
+  # Check conditional limitation
+  parameters_have_conditions = any(parameter.conditions for parameter in experiment.parameters)
+  if parameters_have_conditions ^ experiment.is_conditional:
+    raise ValueError(
+      f"For conditional {cls_name}, need both conditions defined in parameters and conditionals variables"
+      " defined in experiment"
+    )
+  if experiment.is_conditional:
+    if num_solutions and num_solutions > 1:
+      raise ValueError(f"{cls_name} with multiple solutions does not support conditional parameters")
+    if experiment.is_search:
+      raise ValueError(f"All-Constraint {cls_name} does not support conditional parameters")
+    validate_conditionals_for_experiment(experiment)
+
+  # Check feature viability of multitask
+  tasks = experiment.tasks
+  if tasks:
+    if experiment.requires_pareto_frontier_optimization:
+      raise ValueError(f"{cls_name} cannot have both tasks and multiple optimized metrics")
+    if experiment.has_constraint_metrics:
+      raise ValueError(f"{cls_name} cannot have both tasks and constraint metrics")
+    if num_solutions and num_solutions > 1:
+      raise ValueError(f"{cls_name} with multiple solutions cannot be multitask")
+    validate_tasks_for_experiment(experiment, cls_name)
+
+  if experiment.linear_constraints:
+    validate_constraints_for_experiment(experiment)
+    check_constraint_feasibility_for_experiment(experiment)
+
+
+def validate_parameter(parameter):
+  # categorical parameter
```

Comment:
don't need these comments as the `if` statement is pretty clear (same for all the comments in this function). If we want to isolate the concepts more clearly, pull them out into functions.